// @trpc/server/dist/adapters/next.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var TRPCError = require('../TRPCError-fe9113e5.js');
var nodeHTTPRequestHandler = require('../nodeHTTPRequestHandler-d30a4185.js');
require('../resolveHTTPResponse-0a8e47b8.js');
require('../codes-76c38a34.js');
require('../index-ae2b3f4b.js');
require('../transformTRPCResponse-11ac69cc.js');

/* eslint-disable @typescript-eslint/no-explicit-any */ // eslint-disable-next-line @typescript-eslint/ban-ts-comment
function createNextApiHandler(opts) {
    return async (req, res)=>{
        function getPath() {
            if (typeof req.query.trpc === 'string') {
                return req.query.trpc;
            }
            if (Array.isArray(req.query.trpc)) {
                return req.query.trpc.join('/');
            }
            return null;
        }
        const path = getPath();
        if (path === null) {
            const error = opts.router.getErrorShape({
                error: new TRPCError.TRPCError({
                    message: 'Query "trpc" not found - is the file named `[trpc]`.ts or `[...trpc].ts`?',
                    code: 'INTERNAL_SERVER_ERROR'
                }),
                type: 'unknown',
                ctx: undefined,
                path: undefined,
                input: undefined
            });
            res.statusCode = 500;
            res.json({
                id: -1,
                error
            });
            return;
        }
        await nodeHTTPRequestHandler.nodeHTTPRequestHandler({
            ...opts,
            req,
            res,
            path
        });
    };
}

exports.createNextApiHandler = createNextApiHandler;
